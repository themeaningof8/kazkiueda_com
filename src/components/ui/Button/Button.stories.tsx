import type { Meta, StoryObj } from '@storybook/react-vite';
import { Button } from './Button';

const meta: Meta<typeof Button> = {
  title: 'Components/UI/Button', // 階層化命名
  component: Button,
  parameters: {
    layout: 'centered', // レイアウト設定
    docs: {
      description: {
        component:
          'プロジェクトの基本ボタンコンポーネント。Radix UI Slotを使用し、shadcn/uiのデザインシステムに基づいています。',
      },
    },
  },
  argTypes: {
    variant: {
      control: 'select',
      options: [
        'default',
        'destructive',
        'outline',
        'secondary',
        'ghost',
        'link',
      ],
      description: 'ボタンの視覚的バリエーション',
    },
    size: {
      control: 'select',
      options: ['default', 'sm', 'lg', 'icon'],
      description: 'ボタンのサイズ',
    },
    asChild: {
      control: 'boolean',
      description: 'Radix UI Slotとして使用するかどうか',
    },
    disabled: {
      control: 'boolean',
      description: 'ボタンを無効化するかどうか',
    },
  },
  tags: ['autodocs'], // 自動ドキュメント生成
};

export default meta;
type Story = StoryObj<typeof meta>;

// 基本Story
export const Default: Story = {
  args: {
    children: 'ボタン',
    variant: 'default',
  },
};

// 全バリエーション表示
export const AllVariants: Story = {
  render: () => (
    <div className="flex flex-wrap gap-4">
      <Button variant="default">Default</Button>
      <Button variant="destructive">Destructive</Button>
      <Button variant="outline">Outline</Button>
      <Button variant="secondary">Secondary</Button>
      <Button variant="ghost">Ghost</Button>
      <Button variant="link">Link</Button>
    </div>
  ),
  parameters: {
    docs: {
      description: {
        story: 'すべてのボタンバリエーションを表示',
      },
    },
  },
};

// サイズバリエーション
export const AllSizes: Story = {
  render: () => (
    <div className="flex items-center gap-4">
      <Button size="sm">Small</Button>
      <Button size="default">Default</Button>
      <Button size="lg">Large</Button>
      <Button size="icon">🔥</Button>
    </div>
  ),
  parameters: {
    docs: {
      description: {
        story: 'すべてのサイズバリエーションを表示',
      },
    },
  },
};

// インタラクションStory
export const WithInteraction: Story = {
  args: {
    children: 'クリックしてください',
    variant: 'default',
  },
  play: async ({ _canvasElement }) => {
    // インタラクションテストは後で実装
    // const canvas = within(_canvasElement);
    // const button = canvas.getByRole('button');
    // await userEvent.click(button);
    // await expect(button).toHaveClass('active');
  },
  parameters: {
    docs: {
      description: {
        story: 'ボタンのインタラクションテスト用',
      },
    },
  },
};

// 無効化状態
export const Disabled: Story = {
  args: {
    children: '無効化されたボタン',
    variant: 'default',
    disabled: true,
  },
  parameters: {
    docs: {
      description: {
        story: 'ボタンの無効化状態',
      },
    },
  },
};

// アクセシビリティテスト
export const AccessibilityTest: Story = {
  args: {
    children: 'アクセシビリティテスト',
    variant: 'default',
  },
  parameters: {
    docs: {
      description: {
        story: 'アクセシビリティテスト用のボタン',
      },
    },
    a11y: {
      config: {
        rules: [
          { id: 'color-contrast', enabled: true },
          { id: 'keyboard-navigation', enabled: true },
        ],
      },
    },
  },
};

// レスポンシブテスト
export const ResponsiveTest: Story = {
  args: {
    children: 'レスポンシブテスト',
    variant: 'default',
  },
  parameters: {
    docs: {
      description: {
        story: 'レスポンシブ表示確認用',
      },
    },
    viewport: {
      viewports: {
        mobile: { name: 'Mobile', styles: { width: '375px', height: '667px' } },
        tablet: {
          name: 'Tablet',
          styles: { width: '768px', height: '1024px' },
        },
        desktop: {
          name: 'Desktop',
          styles: { width: '1200px', height: '800px' },
        },
      },
    },
  },
};
