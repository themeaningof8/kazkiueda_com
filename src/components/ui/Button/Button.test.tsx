import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { describe, expect, it, vi } from 'vitest'

import { Button } from './Button'

describe('Button', () => {
  it('renders button with text', () => {
    const { asFragment } = render(<Button>Click me</Button>)
    expect(screen.getByRole('button', { name: 'Click me' })).toBeInTheDocument()
    expect(asFragment()).toMatchSnapshot()
  })

  it('handles click events', async () => {
    const handleClick = vi.fn()
    render(<Button onClick={handleClick}>Click me</Button>)
    const button = screen.getByRole('button', { name: 'Click me' })
    await userEvent.click(button)
    expect(handleClick).toHaveBeenCalledTimes(1)
  })

  it('renders as a link when using asChild', () => {
    const { asFragment } = render(
      <Button asChild>
        <a href='/test'>Link Button</a>
      </Button>
    )
    const link = screen.getByRole('link', { name: 'Link Button' })
    expect(link).toBeInTheDocument()
    expect(link).toHaveAttribute('href', '/test')
    expect(asFragment()).toMatchSnapshot()
  })

  it('is disabled when disabled prop is true', () => {
    const { asFragment } = render(<Button disabled>Disabled</Button>)
    const button = screen.getByRole('button', { name: 'Disabled' })
    expect(button).toBeDisabled()
    expect(asFragment()).toMatchSnapshot()
  })

  it('forwards ref correctly', () => {
    const ref = { current: null }
    render(<Button ref={ref}>Button</Button>)
    expect(ref.current).toBeInstanceOf(HTMLButtonElement)
  })

  // Variant and Size Snapshot Tests
  const variants = ['default', 'destructive', 'outline', 'secondary', 'ghost', 'link'] as const
  variants.forEach(variant => {
    it(`renders correctly with variant ${variant}`, () => {
      const { asFragment } = render(<Button variant={variant}>{variant}</Button>)
      expect(asFragment()).toMatchSnapshot()
    })
  })

  const sizes = ['default', 'sm', 'lg', 'icon'] as const
  sizes.forEach(size => {
    it(`renders correctly with size ${size}`, () => {
      const { asFragment } = render(<Button size={size} />)
      expect(asFragment()).toMatchSnapshot()
    })
  })
})
