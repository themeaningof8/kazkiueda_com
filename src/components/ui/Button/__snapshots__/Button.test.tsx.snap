// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > is disabled when disabled prop is true 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3"
    data-slot="button"
    disabled=""
  >
    Disabled
  </button>
</DocumentFragment>
`;

exports[`Button > renders as a link when using asChild 1`] = `
<DocumentFragment>
  <a
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3"
    data-slot="button"
    href="/test"
  >
    Link Button
  </a>
</DocumentFragment>
`;

exports[`Button > renders button with text 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3"
    data-slot="button"
  >
    Click me
  </button>
</DocumentFragment>
`;

exports[`Button > renders correctly with size default 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3"
    data-slot="button"
  />
</DocumentFragment>
`;

exports[`Button > renders correctly with size icon 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 size-9"
    data-slot="button"
  />
</DocumentFragment>
`;

exports[`Button > renders correctly with size lg 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-10 rounded-md px-6 has-[>svg]:px-4"
    data-slot="button"
  />
</DocumentFragment>
`;

exports[`Button > renders correctly with size sm 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-8 rounded-md gap-1.5 px-3 has-[>svg]:px-2.5"
    data-slot="button"
  />
</DocumentFragment>
`;

exports[`Button > renders correctly with variant default 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3"
    data-slot="button"
  >
    default
  </button>
</DocumentFragment>
`;

exports[`Button > renders correctly with variant destructive 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-destructive text-white shadow-xs hover:bg-destructive/90 focus-visible:ring-destructive/20 dark:focus-visible:ring-destructive/40 dark:bg-destructive/60 h-9 px-4 py-2 has-[>svg]:px-3"
    data-slot="button"
  >
    destructive
  </button>
</DocumentFragment>
`;

exports[`Button > renders correctly with variant ghost 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 h-9 px-4 py-2 has-[>svg]:px-3"
    data-slot="button"
  >
    ghost
  </button>
</DocumentFragment>
`;

exports[`Button > renders correctly with variant link 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive text-primary underline-offset-4 hover:underline h-9 px-4 py-2 has-[>svg]:px-3"
    data-slot="button"
  >
    link
  </button>
</DocumentFragment>
`;

exports[`Button > renders correctly with variant outline 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive border bg-background shadow-xs hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 h-9 px-4 py-2 has-[>svg]:px-3"
    data-slot="button"
  >
    outline
  </button>
</DocumentFragment>
`;

exports[`Button > renders correctly with variant secondary 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-secondary text-secondary-foreground shadow-xs hover:bg-secondary/80 h-9 px-4 py-2 has-[>svg]:px-3"
    data-slot="button"
  >
    secondary
  </button>
</DocumentFragment>
`;
